Change the default auth backend to session

norm opened the pull request:
As part of AIP-42, change the default auth backend to validate using the session, so that the UI can use the API. If `auth_backends` has been set to a non-default value, include the session in the list of backends.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

norm committed:
Change default backend

jedcunningham commented on chart/values.yaml:
```diff
@@ -1384,7 +1384,7 @@ config:
   # Authentication backend used for the experimental API
   api:
-    auth_backends: airflow.api.auth.backend.deny_all
+    auth_backends: airflow.api.auth.backend.session
```
This will need to be conditional based on the airflow version.

We might also consider just removing this completely and let the default handle it - the default changed back in 1.10.11, I think it'd be okay as long as we add it to the charts UPDATING.md too.

ashb replied:
Yeah, lets just remove this in the chart values entirely

norm replied:
Removed.

ashb committed:
When updating a deprecated config value from env, set it back to env

Otherwise this means the config seen by an execed sub-process would be
different (and wrong, taking neither the configured env var value, nor
the new default, but instead just what is in the config file!)

uranusjr commented on airflow/configuration.py:
```diff
@@ -350,10 +350,11 @@ def _using_old_value(self, old, current_value):
         env_var = self._env_var_name(section, name)
-        os.environ.pop(env_var, None)
+        # If the config comes from environment, set it there so that any subprocesses keep the same override!
+        if os.environ.get(env_var):
```
```suggestion
        if env_var in os.environ:
```

This matters because it is possible to set an environment variable to an empty string.

ashb committed:
Update airflow/configuration.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

norm committed:
Remove the chart auth_backends setting

norm committed:
Use more accurate warning text when adding session

jedcunningham commented on chart/values.yaml:
```diff
@@ -1382,9 +1382,6 @@ config:
     colored_console_log: 'False'
     remote_logging: '{{- ternary "True" "False" .Values.elasticsearch.enabled }}'
-  # Authentication backend used for the experimental API
-  api:
-    auth_backends: airflow.api.auth.backend.deny_all
```
This looks good, though we should probably add something in `chart/UPDATING.md` about it no longer being set.

jedcunningham replied:
🤦‍♂️ oops, already done!

jedcunningham commented on docs/helm-chart/airflow-configuration.rst:
```diff
@@ -46,7 +46,7 @@ application. See the bottom line of the example:
        remote_logging: '{{- ternary "True" "False" .Values.elasticsearch.enabled }}'
      # Authentication backend used for the experimental API
      api:
```
```suggestion
```

jedcunningham commented on docs/helm-chart/airflow-configuration.rst:
```diff
@@ -46,7 +46,7 @@ application. See the bottom line of the example:
      # Authentication backend used for the experimental API
      api:
-       auth_backends: airflow.api.auth.backend.deny_all
+       auth_backends: airflow.api.auth.backend.session
```
```suggestion
```

jedcunningham replied:
I removed the noise from this example already in another PR.

jedcunningham commented on docs/apache-airflow/security/api.rst:
```diff
@@ -22,12 +22,12 @@ API Authentication
 
     [api]
-    auth_backends = airflow.api.auth.backend.deny_all
+    auth_backends = airflow.api.auth.backend.session
```
Should we add a note to the versionchanged 2.3 section below that it now uses `session` vs `deny_all`?
